### ISSUE TITLE
add id or shortcut for endpoints

### ISSUE THREAD

**Issue by calvix** [giantswarm/gsctl#129]
New endpoint system is cool, but I find it weird that I need to use the whole URL when I want to change to endpoint.

current state:
```
 #: gsctl list endpoint
ENDPOINT URL                                          EMAIL                 SELECTED  LOGGED IN
https://api.g8s.gauss.eu-central-1.aws.gigantic.io    vaclav@giantswarm.io  no        yes
https://api.g8s.ginger.eu-central-1.aws.gigantic.io   vaclav@giantswarm.io  no        yes
https://api-g8s.giantswarm.io                         vaclav@giantswarm.io  no        yes
[calvix@nitro] ~ | 
 #: gsctl select endpoint https://api-g8s.giantswarm.io 
Endpoint selected: https://api-g8s.giantswarm.io
```

I would like something like:
```
 #: gsctl list endpoint
ENDPOINT ID             ENDPOINT URL                                          EMAIL                 SELECTED  LOGGED IN
1                       https://api.g8s.gauss.eu-central-1.aws.gigantic.io    vaclav@giantswarm.io  no        yes
2                       https://api.g8s.ginger.eu-central-1.aws.gigantic.io   vaclav@giantswarm.io  no        yes
3                       https://api-g8s.giantswarm.io                         vaclav@giantswarm.io  no        yes
[calvix@nitro] ~ | 
 #: gsctl select endpoint 3
Endpoint selected: https://api-g8s.giantswarm.io
```

or  with shortcut
```
ENDPOINT ID             ENDPOINT URL                                          EMAIL                 SELECTED  LOGGED IN
gauss                   https://api.g8s.gauss.eu-central-1.aws.gigantic.io    vaclav@giantswarm.io  no        yes
ginger                  https://api.g8s.ginger.eu-central-1.aws.gigantic.io   vaclav@giantswarm.io  no        yes
lycan                   https://api-g8s.giantswarm.io                         vaclav@giantswarm.io  no        yes
[calvix@nitro] ~ | 
 #: gsctl select endpoint lycan
Endpoint selected: https://api-g8s.giantswarm.io
```

Shortcut could be defined when adding endpoint via flag
```
gsctl -e  https://api-g8s.giantswarm.io -n lycan login vaclav@giantswarm.io
```
if not specified we could either provide some random code word from predefined list, assign number or make it required flag

**Comment by calvix** [comment]
WDYT guys @marians @oponder

**Comment by marians** [comment]
I can totally relate to the wish to abbreviate. But I see downsides in all of them. Especially when the user creates an alias for an endpoint, he should also be able to change it at some point. So this requires more tooling for something that should simplify things.

Simply numbering the endpoints has some appeal to me, however I would not use the term "ID" here, as that sounds like something that would be unique and immutable. To me it would be just a number.

```
ENDPOINT NO  ENDPOINT URL                                          EMAIL                 SELECTED  LOGGED IN
1            https://api.g8s.gauss.eu-central-1.aws.gigantic.io    vaclav@giantswarm.io  no        yes
2            https://api.g8s.ginger.eu-central-1.aws.gigantic.io   vaclav@giantswarm.io  no        yes
3            https://api-g8s.giantswarm.io                         vaclav@giantswarm.io  no        yes
```

There is a risk here in the possibility that an endpoint is added or removed and e. g. the number 2 suddenly stands for a different endpoint than before, without the user being aware.

Still, this one is my preferred model of the ones proposed here, as it doesn't involve further maintenance needs.

**Comment by calvix** [comment]
I am fine with numbers.

I guess its specific to our case since a user is not often changing endpoints as we do.

**Comment by calvix** [comment]
We might just increase the `number` counter and dont't reuse deleted numbers

**Comment by oponder** [comment]
what if we tell the user to edit the config file with their favourite text editor if they wish to change the endpoint nickname? that way no further cli tooling would be needed.

**Comment by xh3b4sd** [comment]
I am not fine with numbers. I want to be able to set an identifier that I think makes sense. I personally want to use the codename then. I am fine with editing config files manually when I want to change the codename.

**Comment by marians** [comment]
Another idea: `gsctl select <endpoint>` could support substring matching for the `<endpoint>` argument, as long as that match is not ambiguous.

For example, with these two endpoints configured ...

```
https://api.g8s.gauss.eu-central-1.aws.gigantic.io
https://api.g8s.ginger.eu-central-1.aws.gigantic.io
```

... a user could enter either of the following commands to select the second of the two:

```
gsctl select ginger
gsctl select ginge
gsctl select ging
gsctl select gin
gsctl select ing
```

All of these would work because the string would only match the second, not the first endpoint URL.

**Comment by xh3b4sd** [comment]
I am not fuzzed about the substring matching. When you want more convenience it would be awesome to have shell completion. I am in favour of using the codename because it is the identifier we use in a lot of places. The API endpoint is just one property of the installation which we identify by the codename. The codename is also way more easy to read because it so super short compared to the API endpoint where you have to read carefully where the difference in the string is. When we use the API endpoint as identifier for host clusters we should do the same for guest clusters. I cannot imagine any user would be happy with such an ID. It just does not make sense.

**Comment by marians** [comment]
@xh3b4sd But you understood that my proposal gives you the option to simply do `gsctl select endpoint <codename>`, which is - as I understand - what you want?

**Comment by xh3b4sd** [comment]
I understood that. Thanks.

**Comment by marians** [comment]
Tim and I just talked about this and decided that we (likely: myself) will implement an alias function under the hood. It will allow for an `alias:` key in each endpoint entry in the config.yaml file. If one or more aliases is present, `gsctl list endpoints` will have an `ALIAS` column in the front. The command `gsctl select endpoint <arg>` will allow to use that alias as an argument.

**Comment by teemow** [comment]
Another idea would be to use the installation name itself for an alias. This can be fetched from the `/v4/info` endpoint in each installation and means the user doesn't have to provide any additional arguments. this would also mean the alias will be identical for all users.

btw regarding the substring matching. this can easily be done by your shell if we would provide endpoint completion.

**Comment by marians** [comment]
Please join the discussion on exposing installation code names to customers here: https://github.com/giantswarm/giantswarm/issues/2244

**Comment by marians** [comment]
#180 has been merged. This means that there is basic alias support in `master` now. Please read the details in the PR.

This is still undocumented, as it requires manual editing of the `config.yaml`.

Next step IMHO should be to automatically fetch the installation codename from the API and use that as an alias. Users could still change the alias by manually editing config.yaml.

**Comment by marians** [comment]
This has been released yesterday with 0.10.0. Details in the release description. https://github.com/giantswarm/gsctl/releases/tag/0.10.0
